cpp/src/parquet/column_reader.cc

```diff
@@ -348,6 +367,35 @@ std::shared_ptr<Page> SerializedPageReader::NextPage() {
   return std::shared_ptr<Page>(nullptr);
 }
 
+void SerializedPageReader::DecompressPage(int compressed_len, int uncompressed_len,
+                                          std::shared_ptr<Buffer>& page_buffer) {
+  // Grow the uncompressed buffer if we need to.
+  if (uncompressed_len > static_cast<int>(decompression_buffer_->size())) {
+    PARQUET_THROW_NOT_OK(decompression_buffer_->Resize(uncompressed_len, false));
+  }
+
+  if (current_page_header_.type != format::PageType::DATA_PAGE_V2) {
+    PARQUET_THROW_NOT_OK(
+        decompressor_->Decompress(compressed_len, page_buffer->data(), uncompressed_len,
+                                  decompression_buffer_->mutable_data()));
+  } else {
+    // The levels are not compressed in V2 format
+    const auto& header = current_page_header_.data_page_header_v2;
+    int32_t levels_length =
+        header.repetition_levels_byte_length + header.definition_levels_byte_length;
+    uint8_t* decompressed = decompression_buffer_->mutable_data();
+    memcpy(decompressed, page_buffer->data(), levels_length);
```

Comment:
I wonder if the `memcpy` could be avoided at some point.

Reply:
I think it could be avoided but would require some invasive refactoring to separate extracting the levels and the values from the pages.  I might be wrong but my understanding is that typically the levels are much smaller than the values so I wouldn't expect this `memcpy` to cost too much.